sdks/go/pkg/beam/testing/passert/floats.go

```diff
@@ -26,6 +26,56 @@ import (
 	"github.com/apache/beam/sdks/go/pkg/beam/internal/errors"
 )
 
+// EqualsFloat checks that two PCollections of floats are equal, with each element
+// being within a specified threshold of its corresponding element. Both PCollections
+// are loaded into memory, sorted, and compared element by element.
+func EqualsFloat(s beam.Scope, observed, expected beam.PCollection, threshold float64) {
```

Comment:
We should likely check that both the observed and expected PCollection types are convertable to floats at pipeline construction time.

This is the tricky bit with leaning on "beam.T" in the DoFn's Process Element function. It doesn't validate all the types. As the transform and the code is written, the framework can only validate that the two side input types are identical, but this will pass construction if you provide two PCollections of strings, but will fail at pipeline runtime, which is undesirable.

So we should check it when the pipeline is constructed to fail as soon as possible. This is a guiding principle: If you can turn a Pipeline runtime failure into a Pipeline construction failure, Do it.

eg,  [`beam.ValidateNonCompositeType` ](https://github.com/apache/beam/blob/c9f2321d70cac02aa59bd7ea4877cf27209ae4bc/sdks/go/pkg/beam/validate.go#L38) gets the reflect.Type out of the PCollection, which can then be checked to have a number underneath it or similar.   (ala [validateNonComplexNumber](https://github.com/apache/beam/blob/c9f2321d70cac02aa59bd7ea4877cf27209ae4bc/sdks/go/pkg/beam/transforms/stats/util.go#L54) from the stats transforms package.
(and I just noticed that there's an extra "not" in the comment for that *sigh* )

Reply:
As a second note to make this easier: I recommend putting the main implementation code into a "TryEqualsFloats" function that returns an error, with the "EqualsFloats" function calling that, panicking if the returned error isn't nil.

This lets you test your validation error without complex "defer recover" code, while letting users who accept the runtime panic use the shorter function call, with users who want to handle the error themselves use the "Try" call.

Reply:
Added the validation (and split out into a helper since AllWithinBounds did the same thing anyway.)

Is there a reference for splitting a DoFn into that Try function? I haven't been able to find another example of that design pattern in the SDK and I'm not seeing any examples of pulling the errors out of the ParDo0 call

Reply:
Oh good point there. I would not try and capture errors from ParDo0. It would largely fail if the thresholdFn had any output PCollections, which we can see statically it doesn't, and the additional validation is sufficient to cover other errors it might pull since they are more specific than what is checked by ParDo0 at construction.

The examples are largely the beam.ParDo* functions themselves. There's a single general "TryParDo" that does all the work for the ParDo* functions. See https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/pardo.go#L86

I don't recommend duplicating all of that code or using the TryParDo there, just showing the example.  Another example is TryExternal: https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/external.go#L38

Reply:
Done and unit tested